src/bounds.rs

```diff
@@ -61,25 +80,314 @@ impl Add for Bounds {
 
     fn add(self, rhs: Bounds) -> Bounds {
         Bounds {
-            min: self.min + rhs.min,
+            min: self.min.checked_add(rhs.min).unwrap(),
             max: match (self.max, rhs.max) {
-                (Some(m1), Some(m2)) => Some(m1 + m2),
+                (Some(m1), Some(m2)) => Some(m1.checked_add(m2).unwrap()),
                 _ => None,
             },
         }
     }
 }
 
+impl Sub for Bounds {
+    type Output = Self;
+
+    fn sub(self, rhs: Bounds) -> Bounds {
+        Bounds {
+            min: match rhs.max {
+                Some(m2) => self.min.saturating_sub(m2),
+                None => 0,
+            },
+            max: match self.max {
+                Some(m1) => Some(m1.saturating_sub(rhs.min)),
+                None => None,
+            },
+        }
+    }
+}
+
 impl Mul<Bounds> for Bounds {
     type Output = Self;
 
     fn mul(self, rhs: Bounds) -> Bounds {
         Bounds {
-            min: self.min * rhs.min,
+            min: self.min.checked_mul(rhs.min).unwrap(),
+            max: match (self.max, rhs.max) {
+                (Some(m1), Some(m2)) => Some(m1.checked_mul(m2).unwrap()),
+                _ => None,
+            },
+        }
+    }
+}
+
+impl Div<Bounds> for Bounds {
+    type Output = Self;
+
+    fn div(self, rhs: Bounds) -> Bounds {
+        Bounds {
+            min: match rhs.max {
+                Some(m2) => self.min.checked_div(m2).unwrap(),
+                None => 0,
+            },
+            max: match self.max {
+                Some(m1) => Some(m1 / usize::max(rhs.min, 1)),
+                _ => None,
+            },
+        }
+    }
+}
+
+impl Shl<Bounds> for Bounds {
+    type Output = Self;
+
+    fn shl(self, rhs: Bounds) -> Bounds {
+        Bounds {
+            min: self
+                .min
+                .checked_shl(u32::try_from(rhs.min).unwrap())
+                .unwrap(),
+            max: match (self.max, rhs.max) {
+                (Some(m1), Some(m2)) => Some(m1.checked_shl(u32::try_from(m2).unwrap()).unwrap()),
+                _ => None,
+            },
+        }
+    }
+}
+
+impl Shr<Bounds> for Bounds {
+    type Output = Self;
+
+    fn shr(self, rhs: Bounds) -> Bounds {
+        Bounds {
+            min: match rhs.max {
+                Some(m2) => self.min.checked_shr(u32::try_from(m2).unwrap()).unwrap(),
+                None => 0,
+            },
+            max: match self.max {
+                Some(m1) => Some(m1.checked_shr(u32::try_from(rhs.min).unwrap()).unwrap()),
+                _ => None,
+            },
+        }
+    }
+}
+
+impl BitOr<Bounds> for Bounds {
+    type Output = Self;
```

Comment:
There are definitely better approximations of bounds or/and I think, but not one that I could devise in the time I had available, and since bounds are conservative approximations it seemed fine. We can handle the exact case specially if it turns out to be important, but keeping it simple is fine for now I think.